routing/result_interpretation_test.go

```diff
@@ -176,6 +176,61 @@ var resultTestCases = []resultTestCase{
 			nodeFailure:        &hops[1],
 		},
 	},
+
+	// Tests an invalid onion payload from a final hop. The final hop should
+	// be failed while the proceeding hops are reproed as successes. The
+	// failure is terminal since the receiver can't process our onions.
+	{
+		name:          "fail invalid onion payload final hop",
+		route:         &routeFourHop,
+		failureSrcIdx: 4,
+		failure:       lnwire.NewInvalidOnionPayload(0, 0),
+
+		expectedResult: &interpretedResult{
+			pairResults: map[DirectedNodePair]pairResult{
+				getTestPair(0, 1): {
+					success: true,
+				},
+				getTestPair(1, 2): {
+					success: true,
+				},
+				getTestPair(2, 3): {
+					success: true,
+				},
+			},
+			finalFailureReason: &reasonError,
+			nodeFailure:        &hops[4],
+		},
+	},
+
+	// Tests an invalid onion payload from an intermediate hop. Only the
+	// reporting node should be failed. The failure is non-terminal since we
+	// can still try other paths.
+	{
+		name:          "fail invalid onion payload intermediate",
+		route:         &routeFourHop,
+		failureSrcIdx: 2,
+		failure:       lnwire.NewInvalidOnionPayload(0, 0),
+
+		expectedResult: &interpretedResult{
+			nodeFailure: &hops[2],
+		},
```

Comment:
No successes for preceding hops?

Reply:
fixed, added `failNode` closure for intermediate hops that also reports successes for preceding channels

Reply:
one q: why does `reportIncoming` test that `errorSourceId` is greater than 2, and then only subtract 2?

Reply:
I think it is a bug. For `errorSourceIdx` 2, we seem to be skipping a success marking

Reply:
makes sense, i tried changing the code and our existing unit tests still passed so we should add that as a test case

Reply:
bug fixed and test case added